File: javascript/oneAway.js
/*
One Away
There are three types of edits that can be performed on strings: insert a character,
remove a character, and replace a character. Given two strings, write a function to check
if they are one edit (or zero edits) away.
*/

const oneAway = function (str1, str2) {
    // Return false if length differs by more than 1
    if (Math.abs(str1.length - str2.length) > 1) {
        return false;
    }

    let hash = {};
    let notChar = 0;

    // Add first string to hash table
    for (let i = 0; i < str1.length; i++) {
        hash[str1[i]] = i;
    }

    console.log('hash', hash);

    for (let k = 0; k < str2.length; k++) {
        if (hash[str2[k]] === undefined) {
            notChar++;
            console.log('noChar at k=', k, ': ', notChar)
            if (notChar > 1) {
                return false;
            }
        }
    }
    return true;
}

// Test Cases
console.log(oneAway('pale', 'ple'), 'TRUE');
console.log(oneAway('pale', 'pales'), 'TRUE');
console.log(oneAway('pale', 'bale'), 'TRUE');
console.log(oneAway('pale', 'bake'), 'FALSE');

File: python/oneAway.py
# /*
# One Away
# There are three types of edits that can be performed on strings: insert a character,
# remove a character, and replace a character. Given two strings, write a function to check
# if they are one edit (or zero edits) away.
# */

def one_away(str1, str2):
    # // Return false if length differs by more than 1
    if abs(len(str1) - len(str2) > 1):
        return False

    hash = {}
    not_char = 0

    # // Add first string to hash table
    for i in range(len(str1)):
        print(str1[i])
        hash[str1[i]] = i

    for k in range(len(str2)):
        print(str2[k])
        if str2[k] in hash:
            not_char += 1
            if not_char > 1:
                return False

    return True


# // Test Cases
print(one_away('pale', 'ple'), 'TRUE')
print(one_away('pale', 'pales'), 'TRUE')
print(one_away('pale', 'bale'), 'TRUE')
print(one_away('pale', 'bake'), 'FALSE')

# NOT SOLVED
